From patchwork Thu Mar 18 20:26:57 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Isaku Yamahata <isaku.yamahata@intel.com>
X-Patchwork-Id: 12149225
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT
	autolearn=unavailable autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 480BAC433E6
	for <kvm@archiver.kernel.org>; Thu, 18 Mar 2021 20:30:11 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 11DB164F2A
	for <kvm@archiver.kernel.org>; Thu, 18 Mar 2021 20:30:11 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S231210AbhCRU3g (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Thu, 18 Mar 2021 16:29:36 -0400
Received: from mga04.intel.com ([192.55.52.120]:4540 "EHLO mga04.intel.com"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S230495AbhCRU3S (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 18 Mar 2021 16:29:18 -0400
IronPort-SDR: 
 M4v4oLC0Q2PZdrAAgXc1VKnEnNFD43o/WM0Ji65ly/baar/XmGj82SOmFK5RIk2zdM4ngogvDL
 xqrHRmj78Y4w==
X-IronPort-AV: E=McAfee;i="6000,8403,9927"; a="187405824"
X-IronPort-AV: E=Sophos;i="5.81,259,1610438400";
   d="scan'208";a="187405824"
Received: from fmsmga006.fm.intel.com ([10.253.24.20])
  by fmsmga104.fm.intel.com with ESMTP/TLS/ECDHE-RSA-AES256-GCM-SHA384;
 18 Mar 2021 13:29:17 -0700
IronPort-SDR: 
 /duqs8LGb1pcam6TOOuo19fqDJYUwNc3hTdWoFo9woYvJCgHvkQNd2dWAyB9/xu0nlbiViaHpk
 Q4MW+VaaU5iA==
X-IronPort-AV: E=Sophos;i="5.81,259,1610438400";
   d="scan'208";a="602864941"
Received: from ls.sc.intel.com (HELO localhost) ([143.183.96.54])
  by fmsmga006-auth.fm.intel.com with ESMTP/TLS/ECDHE-RSA-AES256-GCM-SHA384;
 18 Mar 2021 13:29:17 -0700
From: Isaku Yamahata <isaku.yamahata@intel.com>
To: linux-kernel@vger.kernel.org, stable@vger.kernel.org,
        x86@kernel.org, kvm@vger.kernel.org
Cc: brijesh.singh@amd.com, tglx@linutronix.de, bp@alien8.de,
        isaku.yamahata@gmail.com, thomas.lendacky@amd.com,
        Isaku Yamahata <isaku.yamahata@intel.com>,
        "Kirill A . Shutemov" <kirill.shutemov@linux.intel.com>
Subject: [PATCH] X86: __set_clr_pte_enc() miscalculates physical address
Date: Thu, 18 Mar 2021 13:26:57 -0700
Message-Id: 
 <81abbae1657053eccc535c16151f63cd049dcb97.1616098294.git.isaku.yamahata@intel.com>
X-Mailer: git-send-email 2.25.1
MIME-Version: 1.0
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

__set_clr_pte_enc() miscalculates physical address to operate.
pfn is in unit of PG_LEVEL_4K, not PGL_LEVEL_{2M, 1G}.
Shift size to get physical address should be PAGE_SHIFT,
not page_level_shift().

Fixes: dfaaec9033b8 ("x86: Add support for changing memory encryption attribute in early boot")
Reviewed-by: Kirill A. Shutemov <kirill.shutemov@linux.intel.com>
Signed-off-by: Isaku Yamahata <isaku.yamahata@intel.com>
---
 arch/x86/mm/mem_encrypt.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/mm/mem_encrypt.c b/arch/x86/mm/mem_encrypt.c
index 4b01f7dbaf30..ae78cef79980 100644
--- a/arch/x86/mm/mem_encrypt.c
+++ b/arch/x86/mm/mem_encrypt.c
@@ -262,7 +262,7 @@ static void __init __set_clr_pte_enc(pte_t *kpte, int level, bool enc)
 	if (pgprot_val(old_prot) == pgprot_val(new_prot))
 		return;
 
-	pa = pfn << page_level_shift(level);
+	pa = pfn << PAGE_SHIFT;
 	size = page_level_size(level);
 
 	/*
